Improved documentation and readability of MultiMooseEnum class.

pbehne opened the pull request:
Improves the documentation of the `MultiMooseEnum` class. The following attributes/methods are renamed:

`_current` -> `_current_values`
`erase` -> `eraseSetValue`
`push_back` -> `setAdditionalValue`
`getEnum` -> `getSetValueIDs`
`clear` -> `clearSetValues`
`assign` -> `assignValues`
`remove` -> `removeSetValues`
`operator+=` -> `addValidName`
`contains` -> `isValueSet`


Closes #28134.

pbehne committed:
Renamed MultiMooseEnum's _current attribure to _current_values for clarity.

pbehne committed:
Improved documentation of MultiMooseEnum class. Ref #21834.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2327904/) on 27f5978 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28135/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28135/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0797b81f7e4481a0612d672fe053e18bc97e185a`

pbehne commented:
@loganharbour, I just noticed something else: `MultiMooseEnum` overloads the `operator+=` function. This has the effect of adding values to `_current_values`, the actual values of the variable. The class also derives a `push_back` operator from `MooseEnumBase`. This function has the effect of adding values to '_items', the list of valid values the variable can take, but it does not change the value of `_current_values` because this attribute is not defined in the base class.

I find the fact that we have functions `operator+=` and `push_back` that don't do the same thing extremely confusing, especially without more descriptive names. What do you think?

pbehne commented:
Another thing that has bothered me is that we allow duplicate names in a `MultiMooseEnum`. Is there a reason for this, or can I take it out?

GiudGiud commented on framework/include/utils/MultiMooseEnum.h:
```diff
@@ -27,25 +27,27 @@ typedef std::vector<MooseEnumItem>::const_iterator MooseEnumIterator;
+   * values to an object after it has been initialized.
    */
   MultiMooseEnum(std::string names,
                  std::string default_names = "",
```
should we remove the default names default?

I've done that before, thinking MultiMooseEnum("a b c") was setting it to "a b c" when it was setting it to empty

pbehne replied:
I'm sure having that default option in the initialization is super useful. Maybe we should make the parameter names more descriptive. Something like "names" -> "valid_names" and "default_names" -> "initialization_value".

I'll convert to draft and work on further clarity plus consistency between some of these method names.

lindsayad replied:
I thing having two different constructors would be better than having the single constructor with an empty default

GiudGiud commented:
Duplicates are useful 
For exmaple in NS:
inlet_boundaries = 'top bottom front'
inlet_boundary_types = 'fixed-velocity fixed-flux fixed-velocity'

uses duplicated

pbehne committed:
Renamed MultiMooseEnum methods for clarity

The following methods were renamed as follows:

erase -> eraseSetValue
push_back -> setAdditionalValue
getEnum -> getSetValueIDs
clear -> clearSetValues
assign -> assignValues
remove -> removeSetValues

The operator+= method inherited from MooseEnumBase was deleted because “+=“
does not clearly communicate whether the list of possible variable values, the
variable value itself, or both are changed. This method was replaced by one
with the more descriptive name “addValidName”.

pbehne committed:
Removed default value from constructor's initialization_values, added a consturctor with no parameter to specify initialization_values.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2330581/) on 5f642c9 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28135/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28135/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 5bf0d89a341428b1a008ca843cf64d1327b31d7c`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2330677/) on 5f46466 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/28135/site)

This comment will be updated on new commits.

pbehne committed:
Updated push_back to setAdditionalValue in geochemistry.

pbehne committed:
Updated push_back to setAdditionalValue in geochemistry (again).

pbehne commented:
This one is ready for a review. Once everyone is happy with the renames, I will fix the failing apps.

I also thought about renaming `MultiMooseEnum::contains` -> `MultiMooseEnum::isValueSet`, but held off on that because it will require updating a lot of source code. But if we think it will bring a lot more clarity, I am not against making this change.

GiudGiud commented on framework/include/utils/MultiMooseEnum.h:
```diff
@@ -27,29 +27,48 @@ typedef std::vector<MooseEnumItem>::const_iterator MooseEnumIterator;
    * @param allow_out_of_range - determines whether this enumeration will accept values outside of
-   * it's range of
-   *                       defined values.
+   * it's range of initially defined values, allowing for the possiblity to add additional valid
+   * values to an object after it has been initialized.
```
maybe split the allowing into a different sentence for clarity?
```suggestion
   * its range of initially defined values, allowing for the possiblity to add additional valid
   * values to an object after it has been initialized.
```

GiudGiud commented on framework/include/utils/MultiMooseEnum.h:
```diff
@@ -27,29 +27,48 @@ typedef std::vector<MooseEnumItem>::const_iterator MooseEnumIterator;
+                 std::string initialization_values,
+                 bool allow_out_of_range = false);
+  // We need to explicitly define this version so MultiMooseEnum("one two", "one")
+  // doesn't implicitly convert "one" to a bool use use the 2-parameter constructor
```
```suggestion
  // doesn't implicitly convert "one" to a bool and use the 2-parameter constructor
```

GiudGiud commented on framework/include/utils/MultiMooseEnum.h:
```diff
@@ -27,29 +27,48 @@ typedef std::vector<MooseEnumItem>::const_iterator MooseEnumIterator;
+   * any particular value.
+   * @param valid_names - a list of all possible values (names) for this enumeration
+   * @param allow_out_of_range - determines whether this enumeration will accept values outside of
+   * it's range of initially defined values, allowing for the possiblity to add additional valid
```
```suggestion
   * its range of initially defined values, allowing for the possiblity to add additional valid
```

GiudGiud commented on framework/include/utils/MultiMooseEnum.h:
```diff
@@ -99,25 +118,25 @@ class MultiMooseEnum : public MooseEnumBase
-  void push_back(const std::string & names);
-  void push_back(const std::vector<std::string> & names);
-  void push_back(const std::set<std::string> & names);
-  void push_back(const MultiMooseEnum & other_enum);
```
the nice thing with push_back is that it lets you template on a vector OR a multiMooseEnum

pbehne replied:
Do you think it's worth keeping the less clear name? I've just added it back, but deprecated it.

GiudGiud replied:
I could see it being useful for templating. I have not needed it yet

GiudGiud reviewed (commented):
my recommendation would be to keep all the old APIs and add a mooseDeprecated() to them.
this will significantly limit the amount of patching you'll need to do

even if you are up for patching griffin bison and sam, it ll be nice for fully external applications

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2351575/) on 4937ee7 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28135/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28135/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 9a7e0e1d7c085296a28c0506219b649d7779dd75`

pbehne committed:
Deprecated old names for renamed methods, addressed code review.

pbehne committed:
Renamed MultiMooseEnum::contains -> MultiMooseEnum::isValueSet

pbehne committed:
Fixed some unintensional recursive calls.

GiudGiud commented on framework/include/utils/MultiMooseEnum.h:
```diff
@@ -76,9 +96,14 @@ class MultiMooseEnum : public MooseEnumBase
-   * Contains methods for seeing if a value is in the MultiMooseEnum.
-   * @return bool - the truth value indicating whether the value is present
+   * Methods for seeing if a value is set in the MultiMooseEnum.
+   * @return bool - the truth value indicating whether the value is set
```
Add a note saying isValueSet is the same as contains but should be preferred?

GiudGiud commented on framework/include/utils/MultiMooseEnum.h:
```diff
@@ -173,29 +212,42 @@ class MultiMooseEnum : public MooseEnumBase
+  MooseEnumBase & operator+=(const std::string & name);
+  MooseEnumBase & operator+=(const std::initializer_list<std::string> & names);
+
+  // The following replaces operator+= with a more descriptive name
```
like this one

GiudGiud commented on framework/src/utils/MultiMooseEnum.C:
```diff
@@ -57,143 +72,231 @@ MultiMooseEnum::operator!=(const MultiMooseEnum & value) const
 {
-  return std::find_if(_current.begin(),
-                      _current.end(),
+  mooseDeprecated("MultiMooseEnum::contains is deprecated, use MultiMooseEnum::isValueSet");
```
I m not sure we want to disallow the option to template on MultiMooseEnum with the same templates as std containers

pbehne replied:
Would it be acceptable to add a `mooseInfo` statement clarifying that `contains` only operates on values the variable is set to, not possible values that the variable can be set to? We could also only have the info message active for `devel` and `dbg` builds, or maybe `dbg` builds only.

GiudGiud replied:
I dont see why we would want to both:
- enable users to template with MultiMooseEnum just like a std container
- discourage them from doing so because the names aren't great

I think the new names are great and I'd just start using them over the old ones

GiudGiud reviewed (commented):
this looks good

pbehne committed:
Addressed code review.

GiudGiud commented:
Test failures in zapdos unrelated. Modules parallel passed earlier, unrelated too

dschwen commented:
I just stumbled across this and in my opinion the use of `Set` in those method names is confusing. A `MultiMooseEnum` does not necessarily behave like a set. It can be used as a Vector of MooseEnums (i.e. order and cardinality matter!). Also getEnum -> getSetValueIDs makes the APIs of MooseEnum and MultiMooseEnum confusingly different now.
